Add async option to the Reboot method

damdo opened the pull request:
Adds an async option to the Reboot method to schedule reboots asynchronously.
The caller will get an immediate return without it blocking until the reboot actually happened.
This is useful to peform updates from an on-instance supervised process.

Leverages: https://github.com/gokrazy/gokrazy/pull/319

Context: https://github.com/gokrazy/selfupdate/pull/4#discussion_r2116201013

damdo commented:
cc. @stapelberg

stapelberg commented on updater.go:
```diff
@@ -214,6 +214,23 @@ func (t *Target) Reboot(ctx context.Context) error {
 }
 
+// AsyncReboot asynchronously reboots the target, picking up the updated partitions.
+func (t *Target) AsyncReboot(ctx context.Context) error {
```
Or, we could also add a `...RebootOpts` argument to `Reboot`, in case we have other things about rebooting to configure in the future (e.g. the timeouts). 

Kinda like here: https://pkg.go.dev/github.com/gokrazy/rsync@v0.2.9/rsyncd#WithStderr

damdo replied:
Added

stapelberg replied:
I was referring to the functional options pattern, not just a config struct :) https://golang.cafe/blog/golang-functional-options-pattern.html

When specifying multiple config structs (as the current PR allows), it’s not obvious what the merge/override semantics should be. Functional options don’t have this problem here.

damdo replied:
I see, ok updated to follow the suggested pattern, LMK, thanks!

damdo committed:
add async option to Reboot method

Adds an async option to the Reboot method to schedule reboots asynchronously.
The caller will get an immediate return without it blocking until the reboot actually happened.
This is useful to peform updates from an on-instance supervised process.

damdo commented:
Ran a local test for a selfupdate with this and it works:
```
>>>> CHECKING FOR A SUCCESSFUL SELFUPDATE OF THE GOKRAZY INSTANCE <<<<<<<

2025/06/23 09:57:13 Setting desired image for machine "1596c32b95ff4213ab31be56f936d558" to "673b8b4727067583d1cf63022c6cc9ff3637ae0284790b80fab794b75aa88a86"

current version: 673b8b4727067583d1cf63022c6cc9ff3637ae0284790b80fab794b75aa88a86
desired version:
desired SBOM is empty, retrying...

current version: 673b8b4727067583d1cf63022c6cc9ff3637ae0284790b80fab794b75aa88a86
desired version: 673b8b4727067583d1cf63022c6cc9ff3637ae0284790b80fab794b75aa88a86
current and desired version match, success!
test passed in 22 seconds
```

damdo commented:
@stapelberg I've been using this locally for a bit and seems to work well

stapelberg commented:
Thanks! I have also updated the gok CLI with https://github.com/gokrazy/tools/commit/2979dc9b26a9dcb84dece01f1ea4b78adfeacbc2

damdo commented:
Awesome, thanks! :)
